docs/backend-system/core-services/01-index.md

```diff
@@ -65,10 +88,366 @@ createBackendPlugin({
+
+### Using the service
+
+The following example shows how to get a cache client in your `example` backend plugin and `set` and `get` values from the cache.
+
+```ts
+import {
+  coreServices,
+  createBackendPlugin,
+} from '@backstage/backend-plugin-api';
+
+createBackendPlugin({
+  id: 'example',
+  register(env) {
+    env.registerInit({
+      deps: {
+        cache: coreServices.cache,
+      },
+      async init({ cache }) {
+        const { key, value } = { key: 'test:key', value: 'bob' };
+        await cache.set(key, value, { ttl: 1000 });
+
+        // .. some other stuff.
+
+        await cache.get(key); // 'bob'
+      },
+    });
+  },
+});
+```
+
+## Database
+
+Interacting with databases inside your plugin is something that is quite common, and we provide a `PluginDatabaseManager` part of the core services that you can get `knex` client hooked up to your database which is configured in `app-config.yaml`.
+
+If there's no config provided in `backend.database` then you will automatically get a simple in memory `sqlite3` client for your plugin.
+
+These `PluginDatabaseManager`s are scoped per plugin too, so that table names do not conflict across plugins either.
+
+### Using the service
+
+The following example shows how to get a `PluginDatabaseManager` in your `example` backend plugin and get a `client` for interacting with the database and running some migrations from a `migrationsDir` for your plugin.
+
+```ts
+import {
+  coreServices,
+  createBackendPlugin,
+} from '@backstage/backend-plugin-api';
+
+createBackendPlugin({
+  id: 'example',
+  register(env) {
+    env.registerInit({
+      deps: {
+        database: coreServices.database,
+      },
+      async init({ database }) {
+        const client = database.getClient();
+
+        if (!database.migrations?.skip) {
+          await client.migrate.latest({
+            directory: migrationsDir,
+          });
+        }
+      },
+    });
+  },
+});
+```
+
+## Discovery
+
+When building plugins, you might find that you will need to lookup where in fact another plugins `baseUrl`. This could be for example, a `http` route or some `ws` protocol URL. For this we have the `discovery` service that you can query both the internal and external `baseUrl`s given a plugin ID.
+
+### Using the service
+
+The following example shows how to get the `DiscoveryService` in your `example` backend plugin and making a request to both the internal and external `baseUrl`s for the `derp` plugin.
+
+```ts
+import {
+  coreServices,
+  createBackendPlugin,
+} from '@backstage/backend-plugin-api';
+import { fetch } from 'node-fetch';
+
+createBackendPlugin({
+  id: 'example',
+  register(env) {
+    env.registerInit({
+      deps: {
+        discovery: coreServices.discovery,
+      },
+      async init({ discovery }) {
+        const urls = await Promise.all[
+          discovery.getBaseUrl('derp'),
+          discovery.getExternalBaseUrl('derp'),
+        ];
```

Comment:
just make this into two separate await calls into two variables for clarity :) or a single getBaseUrl and a comment saying like "// Or use getExternalBaseUrl" if you ... blah blah